File: examples/footguns/10_alias_mutation/main.c
/*
 * Footgun: Alias Mutation — C
 *
 * In C, any two pointers of the same (or compatible) type can alias the
 * same memory. Mutating memory through one pointer while another pointer
 * holds a "view" of that memory produces unexpected results. The compiler
 * may also misoptimize: if it assumes two pointers don't alias (strict
 * aliasing rule), it may cache a value in a register and never re-read
 * the memory that was modified through the other pointer.
 *
 * The `restrict` keyword promises the compiler two pointers don't alias,
 * enabling optimization — but C does not enforce this promise. Lying with
 * `restrict` is undefined behavior that the compiler may silently exploit.
 */

#include <stdio.h>
#include <stdlib.h>

/* BROKEN: caller passes the same array for both src and dst.
 * memmove handles overlap; memcpy does not. If you call this with
 * aliasing pointers the behavior is undefined for overlapping ranges. */
void unsafe_copy(int *dst, const int *src, int n) {
    for (int i = 0; i < n; i++) {
        dst[i] = src[i]; /* if dst == src or overlaps, this is a data race
                          * between the write on this iteration and a future
                          * read — behavior depends on the overlap direction */
    }
}

/* BROKEN: two pointers into the same array. Modifying via p while
 * reading via q produces surprising output. */
void aliased_pointers(void) {
    int arr[4] = {1, 2, 3, 4};
    int *p = &arr[0]; /* points to arr[0] */
    int *q = &arr[0]; /* also points to arr[0] — same memory! */

    *p = 99;
    printf("*q after *p = 99: %d\n", *q); /* reads 99, not 1 */
    /* A caller who passed q expecting the original value gets a surprise. */
}

/* BROKEN: restrict lie. The compiler may cache src[0] in a register
 * and use it for all iterations, never re-reading after dst modified it.
 * On optimized builds this can produce wrong output. */
void restrict_lie(int * restrict dst, const int * restrict src, int n) {
    /* Caller passes dst == src — violates restrict, UB on -O2 */
    for (int i = 0; i < n; i++) {
        dst[i] = src[i] + 1;
    }
}

int main(void) {
    printf("--- aliased pointers ---\n");
    aliased_pointers();

    printf("--- restrict lie ---\n");
    int arr[4] = {1, 2, 3, 4};
    restrict_lie(arr, arr, 4); /* lie: dst == src */
    for (int i = 0; i < 4; i++) printf("%d ", arr[i]);
    printf("\n");

    return 0;
}

/*
 * Fix: use __restrict__ annotations correctly (and don't lie), or copy
 * src to a local buffer before writing dst, or use memmove for overlapping
 * copies. In practice, aliasing bugs are caught by -fsanitize=undefined
 * (strict-aliasing violations) or by careful code review.
 */

File: examples/footguns/10_alias_mutation/main.go
// Footgun: Alias Mutation — Go
//
// Go slices are reference types: a slice header (pointer, length, capacity)
// is copied by value, but the underlying array is shared. Appending to one
// slice may or may not affect another slice that shares the backing array,
// depending on capacity. This is a common source of subtle bugs.
//
// Go structs are value types: assigning a struct copies it. Modifying the
// copy does not affect the original — the opposite surprise from slices.
//
// Go has no `restrict` keyword and no strict-aliasing rule. The compiler
// does not assume pointers are non-aliasing. This is safer than C's
// strict-aliasing UB but means some optimizations are foregone.

package main

import "fmt"

// SUBTLE: s1 and s2 share a backing array until s2 grows past cap(s1).
// Writes via s2[i] (within s1's capacity) are visible through s1.
func sliceAliasing() {
	s1 := make([]int, 4, 8) // len=4, cap=8
	s2 := s1[:4]             // shares backing array

	s2[0] = 99
	fmt.Printf("s1[0] after s2[0]=99: %d\n", s1[0]) // 99, not 0

	// append within capacity: still shares
	s2 = append(s2, 10) // len becomes 5, still within cap=8
	s2[0] = 42
	fmt.Printf("s1[0] after append+modify: %d\n", s1[0]) // 42

	// append past capacity: new backing array — no longer aliased
	s3 := make([]int, 4, 4) // cap=4
	s4 := append(s3, 0, 0, 0, 0, 0) // forces reallocation
	s4[0] = 999
	fmt.Printf("s3[0] after s4[0]=999 (reallocated): %d\n", s3[0]) // 0, not 999
}

// SURPRISING: struct assignment copies the value. Modify the copy,
// original is unchanged. This is the opposite of the slice aliasing trap.
type Point struct{ X, Y int }

func structCopy() {
	p1 := Point{1, 2}
	p2 := p1 // copy
	p2.X = 99
	fmt.Printf("p1.X after p2.X=99: %d\n", p1.X) // 1, not 99
}

// CORRECT: use pointers explicitly when you want shared mutation.
func sharedMutation() {
	p1 := Point{1, 2}
	p2 := &p1 // pointer — same memory
	p2.X = 99
	fmt.Printf("p1.X after p2.X=99 (pointer): %d\n", p1.X) // 99
}

func main() {
	fmt.Println("--- slice aliasing ---")
	sliceAliasing()

	fmt.Println("--- struct copy (value semantics) ---")
	structCopy()

	fmt.Println("--- shared mutation via pointer ---")
	sharedMutation()
}

// Key insight: Go's aliasing model has two separate surprises:
// - Slices: reference semantics by default (shared backing array)
// - Structs: value semantics by default (copied on assignment)
// The rule of thumb: if you want to share, use a pointer. If you
// append past capacity, you get a new backing array and lose sharing.
// Go's race detector (-race) catches concurrent aliased mutations.
